File: mpikurs_solutions/c/reduce_scatter_vector_ready.c
#include <mpi.h>
#include <math.h>
#include <stdlib.h>
#include <stdio.h>
#include "ritz.h"

/*
  +-------------+----------------+------------------------------------+
  |  parameter  |  input/output  |  description                       |
  +-------------+----------------+------------------------------------+
  |  n          |  input         |  global vectorlength               |
  |  y          |  input         |  partial sum of global vector      |
  |  x          |  output        |  local elements of full vector     |
  +-------------+----------------+------------------------------------+
  |                                                                   |
  |  Purpose of this subroutine is to collect partial result vectors  |
  |  from all processor and add them up to the global result.         |
  |                                                                   |
 +-------------------------------------------------------------------+
*/

void reduce_vector( int n, double *y, double *x ){

   int nproc, myid;
   int firstind[NPROCMAX], recvcounts[NPROCMAX], ip, nl, ia;
   MPI_Status stat;

   MPI_Comm_size(MPI_COMM_WORLD,&nproc);
   MPI_Comm_rank(MPI_COMM_WORLD,&myid); 

   dist_index( n, firstind );
   for (ip = 0 ; ip < nproc ;ip++){
      recvcounts[ip] = firstind[ip+1] - firstind[ip];

   }
   MPI_Reduce_scatter(&y[0],&x[0],recvcounts,MPI_DOUBLE,MPI_SUM, MPI_COMM_WORLD);
}

File: mpikurs_solutions/f/reduce_scatter_vector_ready.f
      subroutine reduce_vector( n, y, x )
c     ===============================
c  all variables should be explicitely declared !
      implicit none 
c
      integer           n
      real*8            y(*), x(*)
c
c  +-------------+----------------+------------------------------------+
c  |  parameter  |  input/output  |  description                       |
c  +-------------+----------------+------------------------------------+
c  |  n          |  input         |  global vectorlength               |
c  |  y          |  input         |  partial sum of global vector      |
c  |  x          |  output        |  local elements of full vector     |
c  +-------------+----------------+------------------------------------+
c  |                                                                   |
c  |  Purpose of this subroutine is to collect partial result vectors  |
c  |  from all processor and add them up to the global result.         |
c  |                                                                   |
c  +-------------------------------------------------------------------+
c  |  ruler                                                            |
c  |5 7 10   15   20   25   30   35   40   45   50   55   60   65   70 | 
c  ++-+--+----+----+----+----+----+----+----+----+----+----+----+----+-+
c
c  parameter constants
      integer           nprocmax
      parameter         (nprocmax = 128)
c  FORTRAN MPI-INCLUDE-file
      include 'mpif.h'
c  local variables
      integer           ierr, nproc, myid, status(MPI_STATUS_SIZE)
      integer           firstind(0:nprocmax), ip, nl, ia
      integer           recvcounts(0:nprocmax)

      call MPI_COMM_SIZE( MPI_COMM_WORLD, nproc, ierr )
      call MPI_COMM_RANK( MPI_COMM_WORLD, myid, ierr )

      call dist_index( n, firstind )
      do ip = 0 , nproc-1
         recvcounts(ip) = firstind(ip+1) - firstind(ip)
      end do
      call MPI_REDUCE_SCATTER( y, x, recvcounts, MPI_DOUBLE_PRECISION, 
     :                    MPI_SUM, MPI_COMM_WORLD, ierr )
 
      return
      end
